### PR TITLE
Introduction to theory of computation

### LINKED ISSUE #817: Introduction to Theory of Computation
### Brief Summary: 
Have you ever wondered why some computer  programs run so efficiently? Computers programs can be viewed as abstracted mathematical models. A good understanding of the theory of computation is required so as to write  efficient computer programs while focusing on capabilities and limitations of computers.
This article will explore the basics of theory of computation and the role it plays in efficiency of computers.

### Key Takeaways: 
What are the 3-5 most important things the reader should understand or be able to do after reading this article?
1.Importance of theory of computation
2. Basics of Automata
3. Computability theory
4. Complexity theory

### References: 
Please list links to any published content/research that you intend to use to support/guide this article. (If none, please indicate 'N/A'.)
N/A

### PR DESCRIPTION
**Pull request by OMONDIELLY**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)

- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.

**NOTE: (1st time contributors - submit ONLY one article to get familiar with the process and style-once published, you can then submit any number of articles.)**
      
#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

-created an Introduction to theory of computation article
-added image header for the article

This fixes #817

### CODE REVIEW

**Comment by OMONDIELLY** [comment]
This fixes #932

**Comment by hectorkambow** [comment]
👋 Thank you for submitting this PR - we will review it ASAP - all PRs go into our queue will be looked over before being approved to be published.

**Comment by OMONDIELLY** [comment]
Looking forward to it being reviewed as soon as possible

**Comment by OMONDIELLY** [comment]
Hi Zolomohan, on the section just below reviewers, I am seeing a place asking still in progress, Convert to draft.. am I required to do anything?

**Comment by zolomohan** [comment]
No. You can convert a PR to a draft if you want to make changes after you have submitted a PR. It will let us know that, the PR is still in progress.

You don't have to worry about it now.

**Comment by hectorkambow** [comment]
plagiarism check done

File: articles/introduction-to-theory-of-computation/index.md [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+- [Basics-of-Automata](#Automata-Basics)
+- [Computability](#Computability)
+- [Complexity](#Complexity)
+- [Conclusion](#Conclusion)
+
+
+### Introduction
+**[Theory of computation]**(https://en.wikipedia.org/wiki/Theory_of_computation) is a branch of Computer Science that is concerned with how problems can be solved  using algorithms and how efficiently they can be solved
+Real world computers perform computations that by nature run like mathematical models to solve problems in systematic ways. 
+The essence of theory of computation is to help develop mathematical and logical models that run efficiently and to a point of halting.Since all machines that implement logic apply TOC, studying TOC gives learners an insight of computer hardware and software limitations
```
Comment (zolomohan) [inline articles/introduction-to-theory-of-computation/index.md:34]: after theory of computation, mark the abbreviation in () like `(TOC)`.

File: articles/introduction-to-theory-of-computation/index.md [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+<!--more-->
+The Theory of Computation is made up of 3 branches i.e
+- Automata Theory
+- Computability Theory
+- Complexity Theory
+
+
+### Basic of Automata Theory
+This branch of theoretical computer science was developed by mathematicians to simplify the logic of computation by using well defined abstract computational devices(models).
+[Automata Theory](https://en.wikipedia.org/wiki/Automata_theory) is basically the study of abstract computational devices and forms a formal framework for design and analysis new types of computing devices such as bio computers and quantum computers.
```
Comment (zolomohan) [inline articles/introduction-to-theory-of-computation/index.md]: `design and analysis new types of computing devices`

after design and analysis, add `of`.

File: articles/introduction-to-theory-of-computation/index.md [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+These models are essential in several areas of computation (applied and theoretical fields). 
+<!--more-->
```
Comment (zolomohan) [inline articles/introduction-to-theory-of-computation/index.md:63]: remove `<!--more-->`. Here and all the places you have used it. This must be added only after the introduction and we will add it. You don't have to worry about it.

File: articles/introduction-to-theory-of-computation/index.md [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+Through automata, we learn how  problems and compute functions are solved by use of automatons. 
+An Automaton is a logical unit of machine that operate singly on an input and follow a defined pattern or configuration so as to produce a desired output.
```
Comment (zolomohan) [inline articles/introduction-to-theory-of-computation/index.md:67]: remove `so as`.

File: articles/introduction-to-theory-of-computation/index.md [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+FA accepts or rejects input strings while reading the strings from left to right.
+**The tuples are –**
++ **Q:** *Finite set of states*
++ **∑:** *Set of input symbols*
++ **q:** *Initial state*
++ **F:** *Set of final states*
++ **δ:** *Transition function*
+
+<!--more-->
+Finite Automata is useful in building text editors / text preprocessors.  FA are poor models of computer in that they can only perform simple computational tasks. Nonetheless, they are essential parts of algorithms and computers.
```
Comment (zolomohan) [inline articles/introduction-to-theory-of-computation/index.md:100]: Add a period of `FA are poor models of computers`. and remove `in that`. `They can only ... `, frame it as a new sentence.

File: articles/introduction-to-theory-of-computation/index.md [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+<!--more-->
+
+- **Context Free Grammars(CFGs)**  – are more powerful abstract models than FA and are essential in Programming Languages and Natural language research work
```
Comment (zolomohan) [inline articles/introduction-to-theory-of-computation/index.md:103]: after essentially, add `used`.

**Review by zolomohan — CHANGES_REQUESTED** [review]
Overall, a good article. 
Although, I noticed a lot of grammatical errors. Please run this article through [Grammarly](https://www.grammarly.com/) to fix these errors.

Also, I've left some suggestions. Ping me when you have implemented them.

**Comment by OMONDIELLY** [comment]
Sure, will do that.

**Comment by OMONDIELLY** [comment]
@zolomohan , I have made the changes you requested

File: articles/introduction-to-theory-of-computation/index.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+The [Computability theory](https://plato.stanford.edu/entries/computability/) defines whether a problem is “solvable” by any abstract machine. Some problems are computable while  others are not.
+
+Computation is done by various computation models depending on the nature of the problem at hand, among these machines being; Turing machine, Finite state machines, and many others.
+
+### Complexity Theory
+This branch is about studying the cost of solving problems while focusing on resources (time & space) needed as the metric.
+
+**Measuring Complexity:**
+
+Involves analysis of the algorithm to determine how much time it takes while solving a problem. As T(n), running time is usually a complex expression when an algorithm runs on large input sizes. This would mean an exponential growth in the running time hence cost.
```
Comment (zolomohan) [inline articles/introduction-to-theory-of-computation/index.md:104]: @OMONDIELLY 

What does T(n) mean here? And why is the running time usually a complex expression when the input size is large?
Isn't there a possibility of algorithms running on O(n) or O(1)?

I think this line is a bit unclear. Can you explain this a bit better?

File: articles/introduction-to-theory-of-computation/index.md [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+### Complexity Theory
+This theoretical computer science branch is all about studying the cost of solving problems while focusing on resources (time & space) needed as the metric.
+The running time of an algorithm varies with the inputs and usually grows with the size of the inputs.
+
+**Measuring Complexity**
+
+It involves analysis of the algorithm to determine how much time and space it takes to solve a problem. This is know as Time complexity and Space complexity respectively.
+
+To evaluate the time and space complexity of an algorithm, a focus is made on the relative rate of growth of the time taken and the space taken to solve the problem as the size of the input grows. We measure an algorithm's time and space requirement as a function of the input size(n). The time complexity and spacce complexity is expressed using the Big Oh(O) notation where only the highest order term in the algebraic expression is considered while ignoring constant values.
```
Comment (zolomohan) [inline articles/introduction-to-theory-of-computation/index.md:107]: Can you add a link to an article that further explains Big Oh notation?

**Review by zolomohan — CHANGES_REQUESTED** [review]

**Comment by OMONDIELLY** [comment]
@Zolomohan, just to clarify something on the last 2  commits, I meant to push "link added " as the commit message  for 
9178845 instead of "updated link".
I wanted to ammend the  commit message but I decided not to.

**Comment by zolomohan** [comment]
> @zolomohan, just to clarify something on the last 2 commits, I meant to push "link added " as the commit message for
> [9178845](https://github.com/section-io/engineering-education/commit/91788451813502de4bd9b7fea966629ce345844b) instead of "updated link".
> I wanted to ammend the commit message but I decided not to.

That's fine @OMONDIELLY.

**Comment by hectorkambow** [comment]
@OMONDIELLY  👍  great job on your 1st article. 
Please be sure to go over any changes made by either myself of our peer reviewer to incorporate any applicable edits to future articles. 

Be sure to check out our [resources page](https://github.com/section-io/engineering-education/blob/master/new_contributors/resources-page.md) with helpful links meant to improve our communitys overall technical writing abilities.

**Comment by OMONDIELLY** [comment]
Thanks..I'll be on the look for any changes and updates

**Pull request merged**